riff-raff/app/persistence/package.scala

```diff
@@ -34,4 +37,17 @@ object `package` {
   implicit class message2MessageDocument(message: Message) {
     def asMessageDocument: MessageDocument = MessageDocument(message)
   }
+
+  implicit class richList[T](list: List[T]) {
+    def distinctOn[N](f: T => N): List[T] = {
+      list.foldLeft((mutable.Set.empty[N], List.empty[T])){ case ((distinctElements, acc), item) =>
+          val element = f(item)
+          if (distinctElements.contains(element)) {
+            distinctElements -> acc
+          } else {
+            (distinctElements + element) -> (item :: acc)
+          }
+      }._2.reverse
```

Comment:
You can avoid that extra traversal with a lazy fold (`foldRight`).

Reply:
I don't think I can as foldRight will remove the duplicates in a different order (the last copy in the list will be retained rather than the first copy).

Reply:
Ah yes indeed, I thought that would not be a problem.

Reply:
It might not be to be fair, I haven't really thought about it. I'll try it and see if it is still working as expected.

Reply:
Yep - that works happily.